As the user I want to see details of an application of a candidate so that I can check the candidate history

kbertalan opened the pull request.

szabob94 committed:
First version of implementation of statehistory.
Next commit will change the whole js and jsp.

szabob94 committed:
state history is showing properly now.

szabob94 committed:
new application state needs it's fields too.

szabob94 committed:
statehistory and buttons are now synchronized

szabob94 committed:
new application state's channel field is fixed

szabob94 committed:
UI changes, user can "modify" the latest state on the UI without POST

szabob94 committed:
Forgot to add the js file

szabob94 committed:
missing dayOfStart field added to db,entity,dto,ui

szabob94 committed:
Modified the db, added state and state_flow tables, added fk-s
Entities needed.

szabob94 committed:
DB rework (maybe) done

szabob94 committed:
Added State flow data to DB.

szabob94 committed:
Fixed the previous commit

kbertalan commented on persistence/src/main/resources/db/migration/data/integration/V1_12_1__AddStatesToStateTable.sql:
```diff
@@ -0,0 +1,14 @@
+INSERT INTO states (id,name) VALUES (1,'newstate');
```
This SQL migration file contains mission critical data, without it Atsy would not work.

In this sense this SQL migration should be placed into the schema folder in order to every environment (production and integration) would be able to built upon it.

szabob94 replied:
fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/3731abffdcebc27e4fb7c9efdaf80d6c45fba90a

kbertalan commented on persistence/src/main/resources/db/migration/data/integration/V1_12_2__AddStateFlow.sql:
```diff
@@ -0,0 +1,64 @@
+INSERT INTO state_flow (from_id,to_id) VALUES (1,2);
```
Like for the state values, this SQL migration script should be placed into the schema folder.

szabob94 replied:
Fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/fe06da0fcae25b73c5f624b733c182f14e75ba48

kbertalan commented on persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StatesRepositoryIT.java:
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.persistence.repositories;
+/*package com.epam.rft.atsy.persistence.repositories;
```
Please maintain the existing test code.
New repositories should have tests as well.
The pull request will not be accepted without tests.

kbertalan commented on service/src/main/java/com/epam/rft/atsy/service/domain/states/AbstractStateDTO.java:
```diff
@@ -1,6 +1,7 @@
 package com.epam.rft.atsy.service.domain.states;
```
I think this DTO should follow the rename of the State -> StateHistory, so it should be AbstractStateHistoryDTO

With the rename the oddity of having a StateDTO in an AbstractStateDTO would vanish.

szabob94 replied:
fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

kbertalan commented on service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java:
```diff
@@ -71,35 +78,40 @@
-        statesRepository.findByApplicationEntityOrderByStateIndexDesc(applicationEntity);
-    List<StateViewDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
+        statesHistoryRepository.findByApplicationEntityOrderByCreationDateDesc(applicationEntity);
+    List<StateViewHistoryDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
```
Also rename the variable.

szabob94 replied:
fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

kbertalan commented on service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java:
```diff
@@ -71,35 +78,40 @@
 
     Assert.notNull(applicationEntity);
-    List<StateEntity>
+    List<StatesHistoryEntity>
```
Also rename the variable.

szabob94 replied:
fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

kbertalan commented on service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java:
```diff
@@ -71,35 +78,40 @@
 
     SimpleDateFormat simpleDateFormat = new SimpleDateFormat(DATE_FORMAT_CONSTANT);
 
     for (int i = 0; i < stateDTOs.size(); i++) {
```
As we don't know which implementation is used for the List interface here, the get(i) operations can have performance penalty. Switch to foreach instead.

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/ApplicationsServiceImplTest.java:
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.service.impl;
+/*package com.epam.rft.atsy.service.impl;
```
Maintain the existing tests.

szabob94 replied:
fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/1f506e2848edeb4fb1ccfe72e2b1906875047d18

kbertalan commented on service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java:
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.service.impl;
+/*package com.epam.rft.atsy.service.impl;
```
Maintain the existing tests.

kbertalan commented on web/src/main/java/com/epam/rft/atsy/web/controllers/ApplicationStateController.java:
```diff
@@ -1,21 +1,49 @@
+        applicationStates =
+        statesService.getStatesByApplicationId(applicationId);
+
+    for (StateViewHistoryDTO stateViewHistoryDTO : applicationStates) {
```
This is the point where it is better to have a new class in the web layer which contains the state history view from the ui perspective.

You can convert the StateHistoryViewDTO instances into the StateHistoryViewXY object and put in the necessary messages.

kbertalan commented on service/src/main/java/com/epam/rft/atsy/service/domain/states/StateViewHistoryDTO.java:
```diff
@@ -14,7 +14,7 @@
 @EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
-public class StateViewDTO extends AbstractStateDTO {
+public class StateViewHistoryDTO extends AbstractStateDTO {
```
Rename to StateHistoryViewDTO

szabob94 replied:
fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

kbertalan commented on web/src/main/webapp/WEB-INF/pages/application_state.jsp:
```diff
@@ -6,26 +6,246 @@
+              <div class="form-group">
+                  <div class="btn-group" role="group">
+                      <c:choose>
+                          <c:when test="${states[0].stateDTO.name == newstate}">
```
Don't use the localized message to find out which section should be displayed. Use the internal name of the state instead.

kbertalan commented on web/src/main/webapp/WEB-INF/pages/application_state.jsp:
```diff
@@ -6,26 +6,246 @@
+          <form class="form-inline pull-right">
+              <c:if test="${states[0].stateDTO.name != reject && states[0].stateDTO.name != accept}">
+              <div class="form-group">
+                  <label class="form-label">Következő lépés:</label>
```
Don't leave hardcoded messages in the jsp-s

kbertalan commented on web/src/main/webapp/WEB-INF/pages/application_state.jsp:
```diff
@@ -6,26 +6,246 @@
+                              <a href="#" class="btn btn-warning"><spring:message code="statehistory.buttons.onHold"/></a>
+                              <a href="#" class="btn btn-danger"><spring:message code="statehistory.buttons.reject"/></a>
+                          </c:when>
+                          <c:when test="${states[0].stateDTO.name == cv}">
```
I thought the StateFlow entity and database table was introduced to be able to provide these sections dynamically.

kbertalan commented:
In addition to the code comments, please add an SQL script which populates one application with state history records, so we could easily check them without the need to manually enter test data into database (while we don't have the proper functionality to create / modify application states)

szabob94 committed:
Moved the V1_12_1__AddStatesToStateTable.sql into the schema.

szabob94 committed:
Moved the V1_12_2__AddStateFlow.sql into the schema.

szabob94 committed:
The refactor commit. Renamed classes and variables.

szabob94 committed:
fixed the ApplicationsServiceImplTest.

szabob94 committed:
Added an SQL script that populates an application with state history records

szabob94 committed:
fixed the ApplicationsRepositoryIT

daergoth committed:
Corrected StatesHistoryRepositoryIT and created StatesRepository

daergoth committed:
Merge branch 'details-of-an-application-of-a-candidate' of https://github.com/epam-debrecen-rft-2015/atsy into details-of-an-application-of-a-candidate

szabob94 committed:
Created the representation class of StateHistoryViewDTO.
Removed hardcoded message from jsp.
Changed localized message tests to a better one in jsp.

daergoth committed:
Fixed StatesHistoryRepositoryIT conflict

daergoth committed:
Rewrote for-loop with List.get() methods to use iterators in StatesServiceImpl

szabob94 committed:
State switcher buttons are coming from db now.

daergoth committed:
Fixed StatesHistoryServiceImplTest and refactored many naming errors

szabob94 committed:
StateFlowRepositoryIT created.
Added some javadoc comment.

szabob94 committed:
javadoc comments for StateFlowServiceImpl.

daergoth committed:
Added StateFlowServiceImplTest

daergoth committed:
Merge branch 'details-of-an-application-of-a-candidate' of https://github.com/epam-debrecen-rft-2015/atsy into details-of-an-application-of-a-candidate

szabob94 committed:
fixed StateFlowRepositoryIT

daergoth committed:
Fixed typo in assert in StateFlowServiceImplTest

daergoth committed:
Merge branch 'details-of-an-application-of-a-candidate' of https://github.com/epam-debrecen-rft-2015/atsy into details-of-an-application-of-a-candidate

kbertalan commented on persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/ApplicationsRepositoryIT.java:
```diff
@@ -86,24 +88,38 @@ public void findByCandidateEntityShouldFindThreeApplicationForCandidateWithThree
+    try {
+      expectedDate = simpleDateFormat.parse("2016-07-26 11:48:55");
+    } catch (ParseException e) {
+      e.printStackTrace();
```
Never use printStackTrace.

In test usually you don't need to handle exceptions, if an exception occurs, than your test case will fail with error.

kbertalan commented on persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StateFlowRepositoryIT.java:
```diff
@@ -0,0 +1,96 @@
+  @Test
+  public void findByFromStateEntityShouldNotFindStateFlowEntityListWhenInvalidFromStateEntityIsGiven() {
+    //Given
+    StatesEntity statesEntity = this.statesRepository.findOne(INVALID_STATE_ID);
```
As I remember this findOne returns null, so why don't you use statesEntity = null instead?

szabob94 committed:
fixed the ApplicationsRepositoryIT and StateFlowRepositoryIT.

daergoth committed:
Merged master into details-of-an-application-of-a-candidate

daergoth committed:
Added @Transactional to service methods in modified or new services
